migrateBuckets/migrateBuckets.js

```diff
@@ -0,0 +1,327 @@
+#!/usr/bin/env node
+'use strict'; // eslint-disable-line strict
+
+const { models, versioning } = require('arsenal');
+const async = require('async');
+const AWS = require('aws-sdk');
+const bucketclient = require('bucketclient');
+const commander = require('commander');
+const fs = require('fs');
+const http = require('http');
+const werelogs = require('werelogs');
+
+werelogs.configure({ level: 'info', dump: 'error' });
+const logger = new werelogs.Logger('s3utils:migrateBuckets')
+    .newRequestLogger();
+const { BucketInfo } = models;
+const versionIdUtils = versioning.VersionID;
+/* expected input JSON structure:
+* {
+*   bucketd: {
+*       bootstrap: <>,
+*   },
+*   https: {
+*       key: <>,
+*       cert: <>,
+*       ca: <>,
+*   }
+* }
+*/
+
+// configurable params
+const ACCESS_KEY = process.env.ACCESS_KEY;
+const SECRET_KEY = process.env.SECRET_KEY;
+const ENDPOINT = process.env.ENDPOINT;
+
+commander
+.version('0.0.1')
+.option('-b, --bucket-name <bucketName>', 'Name of the bucket')
+.option('-s, --src-canonical-id <srcCanonicalId',
+    'Canonical id of source account')
+.option('-d, --dest-canonical-id <destCanonicalId',
+    'Canonical id of destination account')
+.option('-o, --dest-owner-name <destOwnerName>',
+    'Owner display name of destination account')
+.option('-c, --config-path <configPath>', 'Path to metadata config')
+.parse(process.argv);
+
+const {
+    bucketName,
+    srcCanonicalId,
+    destCanonicalId,
+    destOwnerName,
+    configPath,
+} = commander;
+
+if (!bucketName || !srcCanonicalId ||
+!destCanonicalId || !destOwnerName || !configPath) {
+    logger.error('Missing command line parameter');
+    commander.outputHelp();
+    process.exit(1);
+}
+
+let config;
+try {
+    const data = fs.readFileSync(configPath, { encoding: 'utf-8' });
+    config = JSON.parse(data);
+} catch (err) {
+    logger.error(`Could not parse provided config file: ${err.message}`);
+    process.exit(1);
+}
+
+const s3 = new AWS.S3({
+    accessKeyId: ACCESS_KEY,
+    secretAccessKey: SECRET_KEY,
+    region: 'us-east-1',
+    sslEnabled: false,
+    endpoint: ENDPOINT,
+    s3ForcePathStyle: true,
+    apiVersions: { s3: '2006-03-01' },
+    signatureVersion: 'v4',
+    signatureCache: false,
+    httpOptions: {
+        maxRetries: 0,
```

Comment:
Having no retries here mean that the SDK will give up at the first error. This could be an issue if listing large enough buckets, where a single listing error will halt all processing.

We could use a large enough number of retries so that we reduce this risk, and also, making sure that the script is able to recover when restarted if it stopped in the middle of a migration (I think it's the case already, just double checking).

Reply:
yeah, if the script fails at any point, it can be rerun without any problems. but i'll add in a few retries. does 3 seem reasonable?

Reply:
I'm not sure what is the best number, I'd say 10 would be good enough, it needs to be long enough to avoid failures due to transient issues, but short enough that it stops timely if there's a serious issue. AWS SDK uses exponential backoff between retries AFAIK which is desirable.